text_generation/causal_lm/cpp/continuous_batching/library/src/block_manager.hpp

```diff
@@ -141,16 +145,59 @@ class BlockManager {
         OPENVINO_ASSERT(m_block_table.erase(seq_id) == 1);
     }
 
-    bool can_append_slot(SequenceGroup::CPtr seq_group) {
-        // TODO: optimize this HEURISTIC
-        // it assumes that all sequences require new block, but maybe some of them
-        // don't share the same block
-        // let's count actual number of sequences, where last_block_id is the same
-        return seq_group->num_running_seqs() <= m_allocator.num_free_blocks();
+    void free_sequence_partially(size_t seq_id, size_t block_num) {
+        auto block_table = m_block_table[seq_id];
+        OPENVINO_ASSERT(block_table.size() >= block_num);
+        for (size_t idx = 0; idx < block_num; idx++) {
+            m_allocator.free(block_table.back());
+            if (block_table.back()->is_free()) {
+                m_block_table[seq_id].erase(m_block_table[seq_id].end()-1);
+            }
+        }
+        if (m_block_table.size() == 0) {
+            OPENVINO_ASSERT(m_block_table.erase(seq_id) == 1);
+        }
+    }
+
+    bool can_append_slots(SequenceGroup::CPtr seq_group) {
+        return required_blocks_count(seq_group) <= m_allocator.num_free_blocks();
+    }
+
+    size_t required_blocks_count(SequenceGroup::CPtr seq_group) {
+        std::vector<Sequence::CPtr> running_sequences = seq_group->get_running_sequences();
+        std::set<size_t> last_block_ids; // unique last block indices
+        size_t blocks_count= 0; // totat number of needed blocks for sequence group
+
+        for (auto seq: running_sequences) {
+            auto seq_id = seq->get_id();
+            if (m_block_table.find(seq_id) == m_block_table.end()) {
+                // the block table is empty, so we need to allocate the number of blocks equal to number of logical blocks
+                blocks_count += seq_group->get_num_logical_blocks();
+                continue;
+            }
+            auto& block_table = m_block_table[seq_id];
+            size_t num_physical_blocks = block_table.size();
+            size_t needed_blocks_for_current_seq = seq_group->get_num_logical_blocks() - num_physical_blocks;
+            if (num_physical_blocks == 0) {
+                blocks_count += needed_blocks_for_current_seq;
+                continue;
+            }
+
+            // check that last_block_id for this sequence is unique, as sequences with the same last_block_id share the same block
+            KVCacheBlock::Ptr last_block = block_table.back();
+            size_t last_block_id = block_table.back()->get_index();
+            if (last_block_ids.find(last_block_id) == last_block_ids.end()) {
+
+                // if the last_block_id is unique add needed blocks count to the total nomber of needed blocks
+                blocks_count += needed_blocks_for_current_seq;
+                last_block_ids.insert(last_block_id);
+            }
```

Comment:
do we need `else` branch? E.g. if 2 sequences share the same block - how to handle it?

Reply:
Yes, this logic failed on beam search, as it doesn't handle `copy_on_write()` blocks. 
Added logic to add additional required block for each `copy_on_write()` block, please take a look.